### PR TITLE
Feat new dm flow

### PR DESCRIPTION
**Pull request by seavan**
for 1-1 DMs we now have:

1. item in the sidebar appearing "New direct message"
2. no close button
3. no "Go" button - you either click on the username in the list, or you can enter exact username and press Enter
4. no chips - no reason for chips if there is only 1-1 situation
5. counters for Favourites and All
6. when user types text, with timeout = 1000 search occurs and if nothing is found, it's shown as a result underneath
7. TODO: loading indicator?
8. TODO: noGood attribute seems to be not doing anything?

for Rooms, small fixes:
1. can invite via room creation
2. no go button - it was inactive, no handler
3. same type-to-search timeout thing

### CODE REVIEW

**Comment by anri-asaturov** [comment]
yes, noGood is deprecated

File: src/ui/shared-components/UserPicker.jsx [hunk 1]
```diff
@@ -69,7 +70,14 @@ class UserPicker extends React.Component {
...
         const newValLower = newVal.toLocaleLowerCase();
+        if (this.props.limit === 1) {
+            this.query = newValLower.trim();
+            this.searchUsernameTimeout(this.query);
+            return;
+        }
+        this.searchUsernameTimeout(this.query);
         if (newValLower.length > 1 && ', '.includes(newValLower[newValLower.length - 1])) {
             this.query = newValLower.substr(0, newValLower.length - 1).trim();
             this.tryAcceptUsername();
```
Comment (anri-asaturov) [inline src/ui/shared-components/UserPicker.jsx]: One process of looking for username was started in line 80, here it starts second time and will execute handler 2 times: https://github.com/PeerioTechnologies/peerio-desktop/pull/62/files#diff-3381afe59939ea7c796954d51983b576R127  This probably (i don't see how) will not lead to any bugs but is... confusing a bit.

File: src/ui/shared-components/UserPicker.jsx [hunk 2]
```diff
@@ -69,7 +70,14 @@ class UserPicker extends React.Component {
...
 
     handleTextChange = newVal => {
+        this.foundContact = null;
         const newValLower = newVal.toLocaleLowerCase();
+        if (this.props.limit === 1) {
+            this.query = newValLower.trim();
+            this.searchUsernameTimeout(this.query);
+            return;
+        }
+        this.searchUsernameTimeout(this.query);
```
Comment (anri-asaturov) [inline src/ui/shared-components/UserPicker.jsx]: Do we really want to do contact lookup for every keystroke now?

File: src/ui/shared-components/UserPicker.jsx [hunk 3]
```diff
@@ -81,39 +89,71 @@ class UserPicker extends React.Component {
     // Don't use onKeyPress it won't catch backspace
     // Don't use onKeyUp - text change fires earlier
     handleKeyDown = e => {
-        if (e.key === 'Enter' && this.query !== '') this.tryAcceptUsername();
+        if (e.key === 'Enter' && this.query !== '') {
+            // if we are in 1-1 DM selection and user hits enter
+            if (this.props.limit === 1) {
+                const c = this.foundContact || this.searchUsername(this.query);
+                // if we know for sure contact is there, then go to DM immediately
+                if (!c.loading && !c.notFound) {
```
Comment (anri-asaturov) [inline src/ui/shared-components/UserPicker.jsx]: searchUsername() can return null, null ref exception will be thrown here

File: src/ui/chat/ChatView.jsx [hunk 4]
```diff
@@ -0,0 +1,213 @@
+const React = require('react');
```
[the remark this answers is not in the document]
Comment (seavan) [reply src/ui/chat/ChatView.jsx]: nope

File: src/ui/chat/NewChat.jsx [hunk 5]
```diff
@@ -11,8 +11,8 @@ class NewChat extends React.Component {
     @observable waiting = false;
     @observable picker;
 
-    handleChange = (selected) => {
-        if (selected && selected.length) this.handleAccept(selected);
+    componentDidMount() {
+        chatStore.deactivateCurrentChat();
```
[the remark this answers is not in the document]
Comment (seavan) [reply src/ui/chat/NewChat.jsx]: there is no close button anymore. he can only change it by switching to other chat

File: src/ui/shared-components/UserPicker.jsx (hunk 2 above)
Comment (anri-asaturov) [reply src/ui/shared-components/UserPicker.jsx]: And i mean other than 1x1 dm uses of user picker, for 1x1 i see why we need it

Comment (seavan) [reply src/ui/shared-components/UserPicker.jsx]: it's throttled, hence searchUsernameTimeout. no throttle only for 'Enter' keydown.

File: src/ui/shared-components/UserPicker.jsx (hunk 1 above)
Comment (seavan) [reply src/ui/shared-components/UserPicker.jsx]: it won't execute twice, it will clear the current timeout.

other option is to introduce another if.

Comment (anri-asaturov) [reply src/ui/shared-components/UserPicker.jsx]: tryAcceptUsername doesn't use timeouts, but whatever, i don't really see how to fix this and it's not a big deal

File: src/ui/shared-components/UserPicker.jsx (hunk 3 above)
Comment (seavan) [reply src/ui/shared-components/UserPicker.jsx]: ok, if the username is in 'excluded' list only. added a fix in next commit.

File: src/ui/shared-components/UserPicker.jsx (hunk 2 above)
Comment (anri-asaturov) [reply src/ui/shared-components/UserPicker.jsx]: Yes but do other (chip based) component uses need this lookup calls before they type space/comma/enter?

File: src/ui/shared-components/UserPicker.jsx [hunk 6]
```diff
@@ -81,39 +89,71 @@ class UserPicker extends React.Component {
...
+            this.legacyContactError = c.isLegacy;
+            this.showNotFoundError = c.notFound;
+            this.suggestInviteEmailSent = false;
+            this.foundContact = !c.notFound && c;
+        });
+        return c;
+    }
+
+    async tryAcceptUsername() {
+        const c = this.searchUsername(this.query);
```
[the remark this answers is not in the document]
Comment (seavan) [reply src/ui/shared-components/UserPicker.jsx]: they are now reset in `searchUsernameTimeout`. I currently cannot find any flow which would not work - it was tested.

File: src/ui/shared-components/UserPicker.jsx (hunk 2 above)
Comment (seavan) [reply src/ui/shared-components/UserPicker.jsx]: ok, made it so it works the legacy way.

TBD: get new room creation flow and split the component into three things:
- listview
- searchbox
- container which is different for 1-1, rooms and files and sets behaviour for the first two.

File: src/ui/chat/Chat.jsx [hunk 7]
```diff
@@ -170,47 +170,8 @@ class Chat extends React.Component {
...
-                                        onFileShare={this.shareFiles}
-                                    />
-                                </div>
-                        }
-                        {chat ? <ChatSideBar open={Chat.sidebarOpen} onAddParticipants={this.openUserPicker} /> : null}
-                    </div>
-                    {chat && chat.leaving ? <FullCoverLoader show /> : null}
-                </div>
+                {this.props.children}
+                {chat && chat.leaving ? <FullCoverLoader show /> : null}
```
[the remark this answers is not in the document]
Comment (seavan) [reply src/ui/chat/Chat.jsx]: i think it's not a big deal.

**Comment by anri-asaturov** [comment]
btw, there are conflicts in the branch resolve plz. (This is probably the result of me pushing lint fixes earlier, there are not meaningful changes there, just formatting)

**Comment by seavan** [comment]
@anri-asaturov please review

**Pull request merged**